feat: allow multiple origins set per RelyingParty

obroshnij opened the pull request:
As described in the issue https://github.com/cedarcode/webauthn-ruby/issues/428
when building a Relying Party that is supposed to be used by iOS and Android native clients, we need to be able to specify multiple allowed origins per RP (https://developer.android.com/identity/sign-in/credential-manager#verify-origin)
This is yet not supported by this library though is also part of standard specification https://w3c.github.io/webauthn/#sctn-validating-origin

Current workaround is to fallback to creating multiple RP abstractions per expected client platform (Android, iOS, etc) and select one of them based on `User-Agent` header (for example). That is of course troublesome and can actually be avoided. 

This PR allows a relying party to specify either string for origin or array of strings like in the example: 
 
```ruby
# config/initializers/webauthn.rb

WebAuthn.configure do |config|
  # This value needs to match `window.location.origin` evaluated by
  # the User Agent during registration and authentication ceremonies.
  # config.origin = "https://auth.example.com/"
  config.origin = [
    "https://auth.example.com/",
    "android:apk-key-hash:blablablablablalblalla"
  ]

  # Relying Party name for display purposes
  config.rp_name = "Example Inc."
  config.rp_id  = "example.com"
end
```
The only drawback of this approach is that in case an array is used for `config.origin`, - it's impossssible to "guess" relying party by origin (which maybe we should not even do?) thus having array there and no explicitly set `config.rp_id` would result in `RpIdVerificationError` which imo should be expected. 

Hope this PR makes sense. Please let me know should there be any change requests/questions

santiagorodriguez96 commented on lib/webauthn/authenticator_response.rb:
```diff
@@ -82,11 +82,26 @@ def valid_challenge?(expected_challenge)
-      expected_origin && (client_data.origin == expected_origin)
+      return false unless expected_origin
+
+      case expected_origin
+      when Array
+        expected_origin.include?(client_data.origin) # allow multiple origins as per spec
+      else
+        client_data.origin == expected_origin # keep the default behaviour for backwards compatibility
+      end
```
It's kinda confusing IMO for the `origin` attribute to be either a string or a hash – the name kinda implies that it should be just one.

I was thinking that an option could be to accept the `origin` attribute as we historically have but add a new `allowed_origins` config that will take precedence over the `origin` attribute. Then in a future major version we could deprecate the `origin` attribute and leave only the `allowed_origins`.

What do you think?

obroshnij replied:
hi @santiagorodriguez96 , that is a good question. I was also thinking about having `allowed_origins` config but then rejected this idea because it feels redundant to introduce multiple way to assign a value to a single control that we want to have. Also it would probably cause a mental break for people reading documentation trying to understand how `origin` is different from `allowed_origins`. Though, if the intention is to deprecate `origin` config in future versions - we can of course stick to `allowed_origins` long term. I will work on this change in scope of this PR

santiagorodriguez96 commented on spec/webauthn/authenticator_attestation_response_spec.rb:
```diff
@@ -24,28 +24,101 @@
+    end
+  end
+
+  shared_examples "is valid and verifies" do
```
What do you think about changing the name of this shared example to `a valid attestation response` so when included it reads like:

```
it_behaves_like 'a valid attestation response'
```

obroshnij replied:
sure, makes sense to me. Will switch to that

santiagorodriguez96 commented on spec/webauthn/authenticator_attestation_response_spec.rb:
```diff
@@ -54,48 +127,70 @@
+      context "when rp_id is not set explicitly" do
+        include_context "with rp_id set to", nil
+
+        it "verifies" do
+          expect { attestation_response.verify(original_challenge) }.to raise_error(WebAuthn::RpIdVerificationError)
+        end
```
We should change the name of the spec in the cases where it raises an exception, right?

santiagorodriguez96 commented on spec/webauthn/authenticator_attestation_response_spec.rb:
```diff
@@ -54,48 +127,70 @@
+          expect { attestation_response.verify(original_challenge) }.to raise_error(WebAuthn::RpIdVerificationError)
+        end
+
+        it "is valid" do
+          expect(attestation_response.valid?(original_challenge)).to be_falsey
+        end
```
Same here!

santiagorodriguez96 reviewed (commented):
Hey @obroshnij 👋  thank you so much for your time and effort!

I like where this is going!

I will try to give a proper test to this in the next week to see how it works in a real app. I'll let you know once I've done it 🙂 

Then again, thanks!

obroshnij commented:
@santiagorodriguez96 I know you guys are pretty busy, but just so that I can also manage my expectations, is there a chance to get this patch merged any time soon?

santiagorodriguez96 commented on lib/webauthn/relying_party.rb:
```diff
@@ -31,6 +32,7 @@ def initialize(
       @algorithms = algorithms
       @encoding = encoding
       @origin = origin
+      @allowed_origins = allowed_origins.nil? ? [origin] : allowed_origins
```
Is there any reason for setting `allowed_origins` to `origin` if the former is `nil`? It feels to me that I'd leave it `nil` for transparency if there's no need to set it 🤔

santiagorodriguez96 commented on spec/webauthn/authenticator_attestation_response_spec.rb:
```diff
@@ -54,48 +127,70 @@
+    context "when there is a single origin" do
+      include_context "with a single origin"
 
-    let(:attestation_response) do
-      response = seeds[:security_key_direct][:authenticator_attestation_response]
+      let(:origin) { "http://localhost:3000" }
```
Maybe I'm being a little bit nitpicky but to be honest I find these shared examples making it a little bit confusing and less readable – for example we have like three ways of overriding what we are setting inside the context:

- The one highlighted here:

```
      include_context "with a single origin"

      let(:origin) { "http://localhost:3000" }
```

then:

```
include_context "with rp_id set to", "localhost"
```


and

```
      include_context "with multiple origins" do
        let(:allowed_origins) do
          [
            seeds[:android_key_direct][:origin],
            "android:apk-key-hash:blablablablablalblalla",
            "localhost"
          ]
        end
      end
```

It feels to me that, in order to avoid blocking this PR while we attempt to iteratively improve the readability of this file, maybe we can go with the more verbose option of avoiding shared examples and contexts for the moment and then take another pass at this file in an attempt to avoid so much repetition as part of different effort. What do you think?

santiagorodriguez96 commented on lib/webauthn/authenticator_response.rb:
```diff
@@ -25,7 +25,7 @@ def initialize(client_data_json:, relying_party: WebAuthn.configuration.relying_
 
     def verify(expected_challenge, expected_origin = nil, user_verification: nil, rp_id: nil)
-      expected_origin ||= relying_party.origin || raise("Unspecified expected origin")
+      expected_origin ||= relying_party.origin || relying_party.allowed_origins || raise("Unspecified expected origin")
```
What do you think about adding a deprecation warning for the users of the gem using the `origin` attribute? That might help migrating to the using only the `allowed_origins` attribute in the future.

santiagorodriguez96 commented on README.md:
```diff
@@ -106,6 +106,13 @@ WebAuthn.configure do |config|
   # the User Agent during registration and authentication ceremonies.
   config.origin = "https://auth.example.com"
 
+  # For the case when a relying party has multiple origins (e.g. different subdomains, native client sending android:apk-key-hash:... like origins in clientDataJson, etc...), you can set the allowed origins instead of the single origin above
```
What do you think about splitting this into multiple lines like we are doing in the rest of the file? That will make it easier to read as right now you have to scroll horizontally in order to be able to read the whole comment.

santiagorodriguez96 reviewed (commented):
Hey @obroshnij! Thank you for taking the time and the patience for addressing the comments. Sorry that it took me some time to get back to you again. 

I've left a couple more comments in order to keep this moving forward 🙂 

Then again thank you!

santiagorodriguez96 commented on lib/webauthn/relying_party.rb:
```diff
@@ -39,10 +41,18 @@ def initialize(
       @legacy_u2f_appid = legacy_u2f_appid
       self.attestation_root_certificates_finders = attestation_root_certificates_finders
+
+      if allowed_origins.nil? && !origin.nil?
```
I wouldn't check if `allowed_origins` is present or not, as in the hypothetical case that a user of the gem has both `allowed_origins` and `origin`, if we don't throw a warning in that case, once we remove the `origin` method the user's code will start to thrown an error.

santiagorodriguez96 commented on lib/webauthn/relying_party.rb:
```diff
@@ -39,10 +41,18 @@ def initialize(
       @legacy_u2f_appid = legacy_u2f_appid
       self.attestation_root_certificates_finders = attestation_root_certificates_finders
+
+      if allowed_origins.nil? && !origin.nil?
```
If you do:

```ruby
WebAuthn.configure do |config|
  config.origin = "https://auth.example.com"
end
```

as the it's suggested by the documentation, you won't get a warning since [`WebAuthn.configure` initializes an instance of `WebAuthn::Configuration`](https://github.com/cedarcode/webauthn-ruby/blob/b0903000ec431bddd4af653ad87e89064e9bd37c/lib/webauthn/configuration.rb#L8) which [initializes an instance of `RelyingParty`](https://github.com/cedarcode/webauthn-ruby/blob/b0903000ec431bddd4af653ad87e89064e9bd37c/lib/webauthn/configuration.rb#L45) and then [delegates the accessors to the relying party](https://github.com/cedarcode/webauthn-ruby/blob/b0903000ec431bddd4af653ad87e89064e9bd37c/lib/webauthn/configuration.rb#L18-L40).

```
irb(main):002> WebAuthn.configure { |config| config.origin = 'https://auth.example.com' }
=> "https://auth.example.com"
irb(main):004> WebAuthn::RelyingParty.new(origin: 'https://auth.example.com')
DEPRECATION WARNING: `WebAuthn.origin` is deprecated and will be removed in future Please use `WebAuthn.allowed_origins` instead that also allows configuring multiple origins per Relying Party
=>
#<WebAuthn::RelyingParty:0x000000011cd28ad0
 @acceptable_attestation_types=["None", "Self", "Basic", "AttCA", "Basic_or_AttCA", "AnonCA"],
 @algorithms=["ES256", "PS256", "RS256"],
 @allowed_origins=nil,
 @attestation_root_certificates_finders=[],
 @credential_options_timeout=120000,
 @encoding=:base64url,
 @id=nil,
 @legacy_u2f_appid=nil,
 @name=nil,
 @origin="https://auth.example.com",
 @silent_authentication=false,
 @verify_attestation_statement=true>
 ```

obroshnij replied:
I'll take care of that, thanks

santiagorodriguez96 commented on lib/webauthn/authenticator_response.rb:
```diff
@@ -25,15 +25,21 @@ def initialize(client_data_json:, relying_party: WebAuthn.configuration.relying_
       verify_item(:authenticator_data)
-      verify_item(:rp_id, rp_id || rp_id_from_origin(expected_origin))
+
+      # note: we are not trying to guess from 'expected_origin' since it is an array and we can not guess
+      #      which one is the correct one. rp_id should either be explicitly set or guessed from only a single origin
```
Another option too could be that if the array has only one element, extract the `rp_id` from that element. What do you think?

obroshnij replied:
that could be an option too. Though feels a bit opportunistic

obroshnij commented on lib/webauthn/version.rb:
```diff
@@ -1,5 +1,5 @@
 
 module WebAuthn
-  VERSION = "3.1.0"
+  VERSION = "3.2.0"
```
yes, makes sense :)

obroshnij commented:
hey @santiagorodriguez96 thanks a lot for another review. I have addressed the points you brought up and I think it should be ready for the next review

obroshnij commented:
hi @santiagorodriguez96 thanks for running the CI here. I have fixed the rubocop issues now

santiagorodriguez96 commented:
Hi @obroshnij! The PR looks good to me!

I'd want to hold off for merging as my plan was to release this as part of `v3.3` as we have a lot of features coming in `v3.2` too and this one is already a pretty big one 😕 

I'll try to release `v3.2` and `v3.3` in the following days 🙂

obroshnij commented:
> Hi @obroshnij! The PR looks good to me!
> 
> I'd want to hold off for merging as my plan was to release this as part of `v3.3` as we have a lot of features coming in `v3.2` too and this one is already a pretty big one 😕
> 
> I'll try to release `v3.2` and `v3.3` in the following days 🙂

awesome, I'm looking forward to that 🚀

alycit commented:
@santiagorodriguez96 any update here on when this will get merged?  My team is also running up against this issue.

obroshnij commented:
@alycit feel free to grab my fork in the meantime to unblock yourself. I'm trying to keep that one updated

anaclair commented:
> @alycit feel free to grab my fork in the meantime to unblock yourself. I'm trying to keep that one updated

We really appreciate this! 

Would you consider releasing your fork as a gem?

obroshnij commented:
> > @alycit feel free to grab my fork in the meantime to unblock yourself. I'm trying to keep that one updated
> 
> We really appreciate this!
> 
> Would you consider releasing your fork as a gem?

no, sorry. I think you can use it without that. But publishing it as a gem would mean that me or someone would have to support it and update in future. Also I'm afraid it would undermine the amount of work authors of the original gem have invested. 
I would still hope to see this PR released

brauliomartinezlm commented on lib/webauthn/authenticator_response.rb:
```diff
@@ -105,8 +125,16 @@ def valid_user_verified?
+    # Extract RP ID from origin in case rp_id is not provided explicitly
     def rp_id_from_origin(expected_origin)
-      URI.parse(expected_origin).host
+      case expected_origin
+      when Array
+        URI.parse(expected_origin.first).host if expected_origin.size == 1
+      when String
+        URI.parse(expected_origin).host
+      end
```
Related to my comment below, if we transform the origin setter to set allowed_origins, we should just use allowed_origins in across this file and always expect to be an array. In such case, I would only handle this branch of logic:

`URI.parse(expected_origin.first).host if expected_origin.size == 1`

brauliomartinezlm commented on README.md:
```diff
@@ -106,6 +106,17 @@ WebAuthn.configure do |config|
   # the User Agent during registration and authentication ceremonies.
   config.origin = "https://auth.example.com"
 
+  # For the case when a relying party has multiple origins
```
If we're gonna put a deprecation warning and change how this is gonna be used, we actually should directly remove the `config.origin` and only put as a viable option `config.allowed_origins = ["https://auth.example.com"]`. We don't need to be explicit about multiplicity in it. It's easily inferred and we have a whole section talking about RPs instantiation which is still supported.

brauliomartinezlm replied:
In that context, I would just remove the rest for the added comments of this block of changes.

obroshnij replied:
adjusted, thank you

brauliomartinezlm commented:
Hi @obroshnij 👋 

First of all, I really appreciate your work on this PR. Contributions and discussion are always welcome and appreciated ❤️ 

This PR took me a while to process. I apologize for the delay in me writing this comment. While @santiagorodriguez96 quickly interacted with you on it, there was always a design decision in the middle that took me ages (with the little time I can invest into OSS) to get my mind wrapped around it. 

Given the gem's design on Relaying Party instantiation, I always felt this was going against that. I'm still pretty doubtful to be honest.

However, I just had a meeting with @santiagorodriguez96  and we decided to move forward with the change proposed in this PR and how it affects in the API. Webauthn specs is clearly talking about an array of expected origins in the validation step for origin values and that made me reflect in favor of it.

I just left you a few comments that I think are important for us to introduce this change. Again, I do apologize for asking them so down the road.

> Also I'm afraid it would undermine the amount of work authors of the original gem have invested.
Lastly, this comment showed a considerable amount of respect. Very appreciated. Not that common in the OSS realm so it's nice to acknowledge when seen 👍 

Looking forward to unblock folks that are relying on the fork you did and have them back in the official release.

obroshnij commented on lib/webauthn/authenticator_response.rb:
```diff
@@ -83,11 +93,21 @@ def valid_challenge?(expected_challenge)
 
+    # @return [Boolean]
+    # @param [Array<String>] expected_origin
+    # Validate if one of the allowed origins configured for RP is matching the one received from client
```
yes, you're right - removed them

obroshnij commented on lib/webauthn/relying_party.rb:
```diff
@@ -118,5 +120,18 @@ def verify_authentication(
+        )
+      end
+
+      @origin = new_origin
```
thanks @brauliomartinezlm that is actually a good idea imo

obroshnij commented:
hi @brauliomartinezlm 
I appreciate your feedback and already addressed the changes you requested. They all make sense and now I think made the code even more robust. 
Yeah, OSS contributions are sometimes challenging but should definitely be appreciated. This gem is well designed and well thought of and that definitely deserves a 👍 
Happy to move it forward and looking forward to more feedback

brauliomartinezlm commented on lib/webauthn/authenticator_response.rb:
```diff
@@ -25,15 +25,23 @@ def initialize(client_data_json:, relying_party: WebAuthn.configuration.relying_
       verify_item(:authenticator_data)
-      verify_item(:rp_id, rp_id || rp_id_from_origin(expected_origin))
+
+      # NOTE: we are trying to guess from 'expected_origin' only in case it's a single origin
+      # (array that contains a single element)
+      # rp_id should either be explicitly set or guessed from only a single origin
```
I feel I'd remove the comments we added for these methods. You did a good job on making them simple enough so they are easily understandable. While I know it's common practice to comment code we've followed the self documented code and only use comments and annotations for strictly necessary cases. I don't honestly think we're in that scenario here.

brauliomartinezlm commented on README.md:
```diff
@@ -104,7 +104,12 @@ For a Rails application this would go in `config/initializers/webauthn.rb`.
   # This value needs to match `window.location.origin` evaluated by
   # the User Agent during registration and authentication ceremonies.
-  config.origin = "https://auth.example.com"
+  # Multiple origins can be used in more complex scenarios (e.g. different subdomains, native client sending android:apk-key-hash:... etc.)
```
I'd reword this comment to:
```
Multiple origins can be used when needed. Using more than one will imply you MUST configure rp_id explicitely. If you need your credentials to be bound to a single origin but you have more than one tenant, please see [our Advanced Configuration section](https://github.com/cedarcode/webauthn-ruby/blob/master/docs/advanced_configuration.md) instead of adding multiple origins.
```
and remove the empty line below.

obroshnij replied:
awesome, thanks for the hint

brauliomartinezlm commented on README.md:
```diff
@@ -104,7 +104,12 @@ For a Rails application this would go in `config/initializers/webauthn.rb`.
   # This value needs to match `window.location.origin` evaluated by
   # the User Agent during registration and authentication ceremonies.
-  config.origin = "https://auth.example.com"
+  # Multiple origins can be used in more complex scenarios (e.g. different subdomains, native client sending android:apk-key-hash:... etc.)
+
+  config.allowed_origins = [
+    "https://auth.example.com",
+    "android:apk-key-hash:blablablablablalblalla"
+  ]
```
I'd leave this example configuration using one single element (the first one) because I'm assuming that is the majority of our gem users use cases.

brauliomartinezlm reviewed (commented):
Thank you so much @obroshnij for your last comment and addressing the comments and feedback.
Left you another round of comments that are entering the nit picky space but will get the PR ready to get merged.
Again, really appreciate your collaboration 🙏

obroshnij commented:
> Thank you so much @obroshnij for your last comment and addressing the comments and feedback. Left you another round of comments that are entering the nit picky space but will get the PR ready to get merged. Again, really appreciate your collaboration 🙏

Thanks @brauliomartinezlm , that sounds great! Appreciate your collaboration. I have addressed the concerns you mentioned. Looking forward to making this change happen!

obroshnij commented:
> Thank you for the changes @obroshnij ! Last round, very picky from my side but straightforward I hope. After that I'm done, I promise 🙏 😄

Thanks @brauliomartinezlm , thanks for following up on this. I have addressed the CRs. Let me know if there is anything else I can help with

brauliomartinezlm reviewed (approved):
Thank you so much for all the work in this PR @obroshnij . Really appreciate the collaboration ❤️ 
I'll merge this now and we'll cut a release tomorrow.

brauliomartinezlm commented:
Released in [v3.4.0](https://rubygems.org/gems/webauthn/versions/3.4.0)

obroshnij commented:
Thank you @brauliomartinezlm . So happy to see that happening! Would immediately switch to newest version!
